src/x509.c

```diff
@@ -5669,76 +5782,137 @@ static int X509PrintExtensions(WOLFSSL_BIO* bio, WOLFSSL_X509* x509, int indent)
+            ret = WOLFSSL_FAILURE;
+            break;
         }
-        XSNPRINTF(tmp, sizeof(tmp) - 1, "%*s", indent + 4, "");
-        tmp[sizeof(tmp) - 1] = '\0';
         for (i = 0; i < sigSz; i++) {
-            char val[5];
-            int valSz = 5;
+            char val[6];
+            int val_len;
 
             if (i == 0) {
-                XSNPRINTF(val, valSz - 1, "%02x", sig[i]);
+                if ((val_len = XSNPRINTF(val, sizeof(val), "%02x", sig[i]))
+                    >= (int)sizeof(val))
+                {
+                    ret = WOLFSSL_FAILURE;
+                    break;
+                }
             }
             else if (((i % 18) == 0)) {
-                tmp[sizeof(tmp) - 1] = '\0';
-                if (wolfSSL_BIO_write(bio, tmp, (int)XSTRLEN(tmp))
+                if (wolfSSL_BIO_write(bio, tmp, tmp_len)
                         <= 0) {
-                    XFREE(sig, NULL, DYNAMIC_TYPE_TMP_BUFFER);
-                    return WOLFSSL_FAILURE;
+                    ret = WOLFSSL_FAILURE;
+                    break;
+                }
+                if ((tmp_len = XSNPRINTF(tmp, sizeof(tmp), ":\n%*s",
+                                         indent + 4, ""))
+                    >= (int)sizeof(tmp))
+                {
+                    ret = WOLFSSL_FAILURE;
+                    break;
+                }
+                if ((val_len = XSNPRINTF(val, sizeof(val), "%02x", sig[i]))
+                    >= (int)sizeof(val))
+                {
+                    ret = WOLFSSL_FAILURE;
+                    break;
                 }
-                XSNPRINTF(tmp, sizeof(tmp) - 1, ":\n%*s", indent + 4, "");
-                XSNPRINTF(val, valSz - 1, "%02x", sig[i]);
             }
             else {
-                XSNPRINTF(val, valSz - 1, ":%02x", sig[i]);
+                if ((val_len = XSNPRINTF(val, sizeof(val), ":%02x", sig[i]))
+                    >= (int)sizeof(val))
+                {
+                    ret = WOLFSSL_FAILURE;
+                    break;
+                }
+            }
+            if (val_len >= (int)sizeof(tmp) - tmp_len - 1) {
+                ret = WOLFSSL_FAILURE;
+                break;
             }
-            XSTRNCAT(tmp, val, valSz);
+            XMEMCPY(tmp + tmp_len, val, val_len);
+            tmp_len += val_len;
+            tmp[tmp_len] = 0;
         }
-        XFREE(sig, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        if (ret == WOLFSSL_FAILURE)
+            break;
 
         /* print out remaining sig values */
-        if ((i > 0) && (((i - 1) % 18) != 0)) {
-                tmp[sizeof(tmp) - 1] = '\0';
-                if (wolfSSL_BIO_write(bio, tmp, (int)XSTRLEN(tmp))
+        if (tmp_len > 0) {
+                if (wolfSSL_BIO_write(bio, tmp, tmp_len)
                         <= 0) {
-                    return WOLFSSL_FAILURE;
+                    ret = WOLFSSL_FAILURE;
+                    break;
                 }
         }
-    }
+    } while (0);
```

Comment:
With refactoring to remove "return" in favor of setting a ret value and continuing on, lets have periodic ret value checks. This feels like a goto, making the code section a little more complex, but with not using the keyword goto.

Reply:
my kingdom for a `__attribute__((__cleanup__(foo)))`!

refactored as you suggest.  this way is also ugly.  like I said, my kingdom for a cleanup attribute!